shared-libs/cht-datasource/src/local/report.ts

```diff
@@ -0,0 +1,74 @@
+import { LocalDataContext } from './libs/data-context';
+import { fetchAndFilter, getDocById, queryDocsByKey, queryDocsByRange } from './libs/doc';
+import { FreetextQualifier, UuidQualifier } from '../qualifier';
+import { Nullable, Page } from '../libs/core';
+import * as Report from '../report';
+import { Doc } from '../libs/doc';
+import logger from '@medic/logger';
+import { validateCursor } from './libs/core';
+
+/** @internal */
+export namespace v1 {
+  const END_OF_ALPHABET = '\ufff0';
+  const isReport = () => (doc: Nullable<Doc>, uuid?: string): doc is Report.v1.Report => {
+    if (!doc) {
+      if (uuid) {
+        logger.warn(`No report found for identifier [${uuid}].`);
+        return false;
+      }
+    } else if (doc.type !== 'data_record' && !doc.form) {
+      logger.warn(`Document [${doc._id}] is not a valid report.`);
+      return false;
+    }
+
+    return true;
+  };
+
+  /** @internal */
+  export const get = ({ medicDb }: LocalDataContext) => {
+    const getMedicDocById = getDocById(medicDb);
+    return async (identifier: UuidQualifier): Promise<Nullable<Report.v1.Report>> => {
+      const doc = await getMedicDocById(identifier.uuid);
+
+      if (!isReport()(doc, identifier.uuid)) {
+        return null;
+      }
+      return doc;
+    };
+  };
+
+  /** @internal */
+  export const getPage = ({ medicDb }: LocalDataContext) => {
+    const getReportsByFreetext = queryDocsByKey(medicDb, 'medic-client/reports_by_freetext');
+    const getReportsByFreetextRange = queryDocsByRange(medicDb, 'medic-client/reports_by_freetext');
+
+    const getDocsFnForFreetextType = (
+      qualifier: FreetextQualifier
+    ): (limit: number, skip: number) => Promise<Nullable<Doc>[]> => {
+      if (qualifier.freetext.includes(':')) {
```

Comment:
I think the Freetext qualifier should have a function that checks whether this is a `key:value` freetext or not. 

Something like `qualifier.isKeyValue()`

Reply:
In CHT code there is strong bias against using classes/methods in non-Angular code (in the sense that we hardly ever do it).  So, in the cht-datasource API we have stuck with the pattern of using modules/namespaces to organize functions instead of mixing functions with properties on objects.  

In this case, we could add an additional function to the `qualifiers.ts` file to do something like `Qualifier.isKeyedFreetextQualifier(unknown)`.

Reply:
That works for me!